backend/src/apiserver/storage/run_store.go

```diff
@@ -213,17 +214,37 @@ func (s *RunStore) GetRun(runId string) (*model.RunDetail, error) {
 	return runs[0], nil
 }
 
-func (s *RunStore) addMetricsAndResourceReferences(filteredSelectBuilder sq.SelectBuilder) sq.SelectBuilder {
+// Apply func f to every string in a given string slice.
+func Map(vs []string, f func(string) string) []string {
+	vsm := make([]string, len(vs))
+	for i, v := range vs {
+		vsm[i] = f(v)
+	}
+	return vsm
+}
+
+func (s *RunStore) addMetricsAndResourceReferences(filteredSelectBuilder sq.SelectBuilder, opts *list.Options) sq.SelectBuilder {
 	resourceRefConcatQuery := s.db.Concat([]string{`"["`, s.db.GroupConcat("rr.Payload", ","), `"]"`}, "")
+	columnsAfterJoiningResourceReferences := append(
+		Map(runColumns, func(column string) string { return "rd." + column }), // Add prefix "rd." to runColumns
+		resourceRefConcatQuery+" AS refs")
+	if opts != nil && opts.SortByFieldIsRunMetric {
+		columnsAfterJoiningResourceReferences = append(columnsAfterJoiningResourceReferences, "rd."+opts.SortByFieldName)
+	}
 	subQ := sq.
-		Select("rd.*", resourceRefConcatQuery+" AS refs").
+		Select(columnsAfterJoiningResourceReferences...).
 		FromSelect(filteredSelectBuilder, "rd").
 		LeftJoin("resource_references AS rr ON rr.ResourceType='Run' AND rd.UUID=rr.ResourceUUID").
 		GroupBy("rd.UUID")
 
+	// TODO(jingzhang36): address the case where some runs don't have the metric used in order by.
 	metricConcatQuery := s.db.Concat([]string{`"["`, s.db.GroupConcat("rm.Payload", ","), `"]"`}, "")
+	columnsAfterJoiningRunMetrics := append(
+		Map(runColumns, func(column string) string { return "subq." + column }), // Add prefix "subq." to runColumns
+		"subq.refs",
+		metricConcatQuery+" AS metrics")
 	return sq.
-		Select("subq.*", metricConcatQuery+" AS metrics").
+		Select(columnsAfterJoiningRunMetrics...).
```

Comment:
Can you write an example of how this will be constructed as a comment?
I don't feel this part can easily be understood by the next person looking at it.

Reply:
Yeah, totally agree the resulting sql becomes really difficult to read. The options we have are
1. re-modeling the way run_details, run_metrics and resource_references are designed since the current design induces table joins when retrieve runs with full information, and does impact the query performance as well. However, this might be very likely a breaking change and a big change too.
2. instead of writing a single query to retrieve runs, we can do the retrieval in several separate queries, and create views for the result of the intermediate queries and use those views in the subsequent queries. This would not improve query performance but does make the queries easier to read.

The above being said, as for now, I can add some comment for the resulting single query to explain what each part of the query is doing.

Reply:
I'll add the comment to the subsequent pr at #4251

Reply:
Thanks for adding the comment!